### ISSUE TITLE
Logical structure, PDF objects as content items (14.7.5): Parent structure elements and object references

### ISSUE THREAD

**Issue by stechio** [pdf-association/pdf-issues#343]
Here I am reporting a possible bug to correct (SUB-ISSUE 1), and requesting a clarification on an apparent contradiction related to it (SUB-ISSUE 2).

### SUB-ISSUE 1: Cardinality of parent structure elements of PDF objects as content items

Describing the structural parent tree from the point of view of a **PDF object as content item** (such as image XObject), subclause 14.7.5.4 (Finding structure elements from content items) states:
> The tree shall contain an entry for each object that is a content item of at least one structure element [...].

To my understanding, *the phrase "at least one" is misleading, since a PDF object that is a content item is required (see also Table 359 (Additional dictionary entries for structure element access)) to be associated to ONE AND ONLY ONE parent structure element*, as later stated:
> The key for each entry shall be an integer given as the value of the StructParent [...] entry in the object [...]. [...] the value [for each entry] shall be an indirect reference to the parent structure element.

This statement is unequivocal, especially considering that, on the other hand, for a content stream (page object or form XObject) containing marked-content sequences that are content items "the value shall be an array of indirect references to the sequences' parent structure elements."

### SUB-ISSUE 2: Parent structure element of a PDF object as content item in case of multiple, separate object references

All the above held, the following statement in NOTE 2 of subclause 14.7.5.3 (PDF objects as content items) depicts a more complex case which, to my understanding, leads to a contradiction:
> If the referenced object is rendered on multiple pages, each rendering requires a separate object reference.

As you know, a PDF object as content item is included in the structure tree (see 14.7.2 (Structure hierarchy)) as an **object reference** (see 14.7.5.3 (PDF objects as content items)). Given that a PDF object as content item can be associated, via structural parent tree, to ONE AND ONLY ONE parent structure element, NOTE 2 implies that its single parent structure element will have to contain ALL its separate object references, each one for the distinct page where it is rendered (am I right?). *However, such conclusion seems to disrupt the whole purpose of logical content order (the ordering for semantic purposes)* as described by subclause 14.8.2.5 (Page content order and logical content order)... What's wrong with it?

What is the **correct way to place in the logical structure multiple, separate object references (one per page) of the same PDF object as content item**? (EXAMPLE 1 in 14.7.5.4 (Finding structure elements from content items) only shows the simple case of a single object reference; IMO, *that example should be expanded to address also the scenario depicted in NOTE 2, ie multiple separate object references, one per page*)

**Comment by petervwyatt** [comment]
@DuffJohnson? @mrbhardy?

**Comment by myang-apryse** [comment]
I ran into this problem when reading the spec as well.
my 2 cents is that, I see a few different consistent resolutions:

1. disallowing multiple OBJRs from referencing the same XObj
    - we basically have to replace & negate the note:
      >  "The tree shall contain an entry for each object that is a content item of at least one structure element"
    - in the case you want to do multiple reference, you'll have to convert your OBJRs to MCRs 
    and tag the "Do" operators at each instance

2. allowing and requiring the removal of StructParent in the Obj itself in this case (potentially a bit hacky)
    - the situation must be that the XObj is tagged via the Do operator in each including stream, and has an MCID in each
    - forward lookup from structure tree will go directly to the Obj via the OBJR in the leaf
      - works the same as single OBJR
    - backward lookup from the content will find that StructParent does not exist, and use instead the StructParents entry of the "owning stream" (this assumes the pdf processor somehow has a context stack to look up)
      - this proceeds as normal using the MCID given by the tagged 'Do' operator

3. allowing and requiring StructParent (singular) to be an Array in this case
    - this will provide enough information solve the multiple reference problem, 
    but will require additional logic to distinguish the instances
      - i.e. will have to look at all the potential linked OBJRs and match to the current context (i.e. via Pg?)
    - but may introduce unintended additional complexity, as all lookups will have potentially to take it into account
      - maybe fine if it's restricted to the XObjects special case? (i.e. not allowed for Pages)

Notes:
- MCRs won't work for Annotations, since those are not referenced in the streams,
but it's unclear reusing annotations is a valid use case, much less tagging them if reused.
- Separately (unrelated to OBJR):
We may want to add additional clarification/constraints on the requirement of fields such as Pg, Stm, StmOwn in the MCR to support unambiguous lookup in cases of multiple reference

**Comment by myang-apryse** [comment]
@stechio 
My interpretation of sub-issue-1 differs from your wrt to intent,
although maybe the spec should be written in a way to avoid this confusion.

My interpretation is that:
The [Number]Tree shall contain an **entry** [that is of type Ref or Array],
for each [PDF] object that is a **[or contains]** content item[s] of at least one structure element.

which leads to 2 situations (which may bear enumerating instead of aggregating into the single sentence).
- is a content item -> StructParent (singular) entry ->ref in ParentTree
  - this leads to sub-issue-2 in the case of multiple OBJR under different parents for the same XObj
  - points directly to a SINGLE parent struct element,
  the cardinality of this is EXACTLY ONE, as you say
- is a container for content items -> StructParents (plural) -> array of refs in ParentTree
  - each entry in the array is a reference which may point at different structure elements,
  we don't want empty array, so the cardinality is ONE OR MORE.

The union of these 2 cases gives the requirement of _at least one_

I believe the bolded section is what caused this confusion,
"entry" is ambiguous/confusing whether it refers to:
- a NumberTree entry
  - in this case we're missing wording that allows for "containment" of content
- an end element reference (I think this may be your interpretation?):
  - whether it be a NumberTree entry itself,
  - or an element within the Array that is the NumberTree entry

**Comment by myang-apryse** [comment]
After a second reading, I agree with your interpretation.

**Comment by stechio** [comment]
@myang-apryse:

> After a second reading, I agree with your interpretation.

However, I appreciated [your resolution proposals for SUB-ISSUE 2](https://github.com/pdf-association/pdf-issues/issues/343#issuecomment-1741493217) (IMHO, they are worth a more in-depth evaluation):

> I ran into this problem when reading the spec as well. my 2 cents is that, I see a few different consistent resolutions:
> 
> 1. disallowing multiple OBJRs from referencing the same XObj
> [...]
> 
> 2. allowing and requiring the removal of StructParent in the Obj itself in this case (potentially a bit hacky)
> [...]
> 
> 3. allowing and requiring StructParent (singular) to be an Array in this case
> [...]

**Comment by petervwyatt** [comment]
@mrbhardy - could you please contribute...

**Comment by car222222** [comment]
Here are some relatively simple ideas for clarifying the relevant current provisions. 
I hope they are useful.

﻿Re: Some Normative stuff that is incorrect/insufficient.

1.
The start of 14.7.5.3 needs clarification, it should perhaps say something like this:
“When a structure element’s content contains an entire PDF object (that is not also contained in the content of any other structure element), such as . . . “

But this restriction on the use of OBJRs could be in a separate sentence. 

2. 
(The wording here is probably the worst problem!)
﻿In 14.7.5.4 there is a misleading phrase, “at least one”
 (since this can never be “more than one”); 
it could replaced by “a”, without at all changing the meaning. 
Suggested change:
“The tree shall contain an entry for each object that is a content item of 
at least one structure element “ 
==>
"The tree shall contain an entry for each dictionary object that is a content item of 
a structure element".
Note that it is very important to alter the phrasing here; this change will not affect the normative meaning. 

In general, it is important to clarify this restriction, with a clear statement such as:
“﻿A PDF object can never appear (as a content item referenced  
by an OBJR) as an item in more than one structure element.” 
This explanation/requirement could be in a Note. 

Re:  Some related, non-normative, texts also need attention. 
This is because some sentences either do not make sense, are not 
sufficiently precise, or they give a wrong impression. 
These following two, in particular, are also from 14.7.5.3, so they are 
closely related to 1. above. 

A.
In Note 1, above Table 356, the second sentence needs a lot of clarification. 
I am not sure what might have been its intended meaning, as it appears to be about 
the content stream of some object, without saying what object this is.  
It would probably be more useful if it said something about the content 
items of a structure element.
Also, the first sentence therein may be clearer as follows:
“This form of reference can only be used for entire objects that are in the content of a 
structure element.”

B.
Note 2 may be the main cause of the current confusion? 
It definitely needs some additions, in order to explain that: 
“An object that is referenced by an OBJR in the content of a structure element 
can be referenced more than once (possibly using a distinct OBJR), but only
when any further references to this object are within the content of that same 
structure element.”
And/or: 
“Each object that is so referenced can appear only in the content of this one 
structure element, and in no others.”

**Comment by stechio** [comment]
@car222222:
> Here are some relatively simple ideas for clarifying the relevant current provisions. I hope they are useful.
> 
> ﻿Re: Some Normative stuff that is incorrect/insufficient.

Could you please provide us some context (your text looks a bit sketchy, as if it was extracted from a mailing-list thread I couldn't trace back)?

**Comment by car222222** [comment]
@stechio 

This was intended to help @mrbhardy to respond 
in a reasonably minimal way to the above (non-specific) question from @petervwyatt. 

It was not derived from any other correspondence.

Naturally, it assumes familiarity with the current documentation of OBJRs and related ideas.
It suggests only how to extend the current documentation so that it accurately and clearly describes the current (normative) situation.
   
It is not intended to deal with, or even discuss,  your concerns about these currently mandated provisions.

**Comment by sgaither** [comment]
SFAIK, the sub-issue #1 should be clarified to indicate only a single parent (i.e., I agree that an MCID should only have a single parent). In the case of sub-issue #2, there are two mechanisms for tagging an XObject container: (1) the contents of XObject container is itself is tagged, but then the Form XObject cannot have a structural element parent (but can be referenced on the page without an associated MCID) or (2) the XObject container has no internal tagging but may be contained in the structure tree exactly once (i.e., a single parent).  That said, in case (2) above, I believe that an inner Form XObject could potentially be contained by multiple outer Form XObjects, each specifying a unique parent in the structure tree, and thereby still allowing for the concept of reuse of content, but not violating the concept of a single parent.

**Comment by myang-apryse** [comment]
Additionally, to clarify, I (now) think there is no substantive difference between StructParent singular and plural.
That is, if you treat the look up as a function from a `pair (SP, MC_IDX) -> indirect reference to structure element node`,
where SP is StructParent(s), and MC_IDX is the MCID, or -1 (n/a) for singular StructParent.

> I agree that an MCID should only have a single parent

Then the above quote should also mean that XObjs, through either method, should have 1 parent structure element node.
The reason is that you no longer have a function if you wanted to refer to multiple "parents", and as I mentioned before, the spirit of the lookup seems to eschew additional context.

**Comment by myang-apryse** [comment]
@sgaither I think I agree with what you're saying wrt XObj2 contained in XObj1, 
as long as XObj2 is not itself tagged through methods 1 & 2, can be referred to multiple times.

And additionally, there's what can be considered a 3rd way to tagging (i.e. surround with structure scope) content inside an XObj, and that is to surround the "do" operator in the parent stream with an MC.
This way, you can use the StructParent(s) of the parent stream,  allowing for proper lookup of multiple parent structure elements from different usages of the same XObj.

In some sense, why this works is the same reason as what you said, 
if you treat case (2) as surrounding the entire obj stream (which contains the "do" for XObj2) with an MC.

**Comment by mrbhardy** [comment]
The PDF/UA TWG agrees with sub-issue 1 and that it should be:

> The tree shall contain an entry for each object that is a content item of a structure element

**Comment by mrbhardy** [comment]
NOTE: @sgaither to provide some examples

**Comment by petervwyatt** [comment]
Waiting for examples from @sgaither before finalizing and closing this issue.

**Comment by jewelblue** [comment]
Upon more careful examination of the specification, the verbiage at the end of 14.7.5.2 is quite clear that:
"A form XObject that is painted with multiple invocations of the **Do** operator shall be incorporated into the Document's logical structure _only by the first method_, with each invocation of **Do** individually associated with a structure element"

First, I propose to append to this final sentence above:
"See Example 6 in this subclause for multiple invocation of a form XObject to overprint a phase in conjunction with the use of **ActualText** to indicate that the phrase is to be read only once (see 14.9.4, "Replacement text"). Note that since each invocation of the **Do** operator is required to be individually associated with a structural element, a single Span directly contained inside the page content stream cannot be used to contain both invocations of the form XObject."

Second, I propose to add the following text contained within the attached file as "EXAMPLE 6":
[32000-2-2020-14.7.5.2-Example-6.txt](https://github.com/pdf-association/pdf-issues/files/15339614/32000-2-2020-14.7.5.2-Example-6.txt)

**Comment by jewelblue** [comment]
(I apologize for using the wrong login on the prior response)
Upon more careful examination of the specification, the verbiage at the end of 14.7.5.2 is quite clear that:
"A form XObject that is painted with multiple invocations of the Do operator shall be incorporated into the Document's logical structure only by the first method, with each invocation of Do individually associated with a structure element"

First, I propose to append to this final sentence above:
"See Example 6 in this subclause for multiple invocation of a form XObject to overprint a phase in conjunction with the use of ActualText to indicate that the phrase is to be read only once (see 14.9.4, "Replacement text"). Note that since each invocation of the Do operator is required to be individually associated with a structural element, a single Span directly contained inside the page content stream cannot be used to contain both invocations of the form XObject."

Second, I propose to add the following text contained within the attached file as "EXAMPLE 6":
[32000-2-2020-14.7.5.2-Example-6.txt](https://github.com/pdf-association/pdf-issues/files/15379952/32000-2-2020-14.7.5.2-Example-6.txt)

**Comment by sgaither** [comment]
Upon more careful examination of the specification, the verbiage at the end of 14.7.5.2 is quite clear that:
"A form XObject that is painted with multiple invocations of the Do operator shall be incorporated into the Document's logical structure only by the first method, with each invocation of Do individually associated with a structure element"

First, I propose to append to this final sentence above:
"See Example 6 in this subclause for multiple invocation of a form XObject to overprint a phase in conjunction with the use of ActualText to indicate that the phrase is to be read only once (see 14.9.4, "Replacement text"). Note that since each invocation of the Do operator is required to be individually associated with a structural element, a single Span directly contained inside the page content stream cannot be used to contain both invocations of the form XObject."

Second, I propose to add the following text contained within the attached file as "EXAMPLE 6":
[32000-2-2020-14.7.5.2-Example-6.txt](https://github.com/pdf-association/pdf-issues/files/15379970/32000-2-2020-14.7.5.2-Example-6.txt)

**Comment by petervwyatt** [comment]
PDF TWG agree but want a few days to review example in detail.

**Comment by robinwatts** [comment]
I'm coming late to this discussion, so please bear with me if I'm asking silly questions.

First off, in the example, shouldn't the /P tag of object 2 be "1 0 R" rather than elipsis? The parent element is included within this example, so should be shown unambigiously.

Secondly, I'm unusure about how "A form XObject that is painted with multiple invocations of the Do operator shall be incorporated into the Document's logical structure only by the first method, with each invocation of Do individually associated with a structure element" is supposed to work at all.

My understanding is that the semantic content of the example given should only contain "clear emphasis" once. It's invoked and hence drawn twice, but a text extraction (or a screen read etc) of the document should only actually contain that once.

I'm trying to understand how I'm supposed to know that /Fm5 should only be counted semantically once.

Is the interpreter supposed to keep a list of all the form XObjects it's encountered so far whilst processing a page stream, and to check every new invocation against that list? What if the same XObject is used from multiple pages? Or on the same page from different annotations?

I could kind of understand how to do this umambiguously if the XObject contained some piece of information that said "I only count semantically when I'm invoked as MCID x", but I can't see anything like that.

I'd claim that this kind of example (same content, multiple uses, only counting once) is MUCH less common than the usual way I'd expect to see this kind of content used (same content, multiple uses, counting every time). For instance, imagine that I've got a document advertising the benefits of new WhizzyShine. Every time I mention WhizzyShine in the text, I want to use the funky WhizzyShine logo, so I encapsulate that into an XObject and reuse that multiple times across multiple pages. During a semantic extraction from the document, I want that to appear every single time.

But this seems to be exactly the thing that we're prohibiting here? There is probably something obvious I'm missing here...

If so, an example that differentiated the 2 different cases would probably help other bears of tiny brain here.

**Comment by myang-apryse** [comment]
@robinwatts, the example shows the contents of multiple distinct pdf objects,
obj 2 is not a stream containing a /P tag, but rather the structure element node,
the "/P" there refers to the P key of the dictionary which points at the parent node in the the tree.

4 0 obj is the actual page stream, and if you look at the contents (the stuff following the stream KEYWORD),
the /Span <</MCID 1>> BDC /Fm5 Do EMC is the actual sequence of instructions that "invokes Fm5".

The instructions are postfix, and you'll have to read the definitions of each in the spec,
but specifically, the above sequence means:
```
BeginMarkedContent(type="Span", properties={"MCID, 1}), // BDC
Do(xobj="Fm5"), // Basically "paste Fm5 here"
EndMarkedContent(), // EMC
```
Note that according to the spec, the "type" in the BDC,
"has no relation to" the "Structure type" of structure element (denoted by the "S" key: `/S Span`)
with the latter being the important one, but they are recommended to be the same (hence perhaps your confusion).

--- 

Wrt your comments on MCID and invoking, I hope the following helps, 
as I think I was confused in a similar way when I first read it, but the truth becomes clear as you work with PDF objs.

The concept of "XObj invocation" is the usage of the "Do" operator, 
and it acts as if you replace the instance of the "Do" instruction with the contents of the XObj stream,
with some caveats on MCID as I'll discuss later:

MCID is a number that is stream local, that is to say, both the page stream and Fm5 xobj stream can contain say MCID 1.
(So your comments about being used on multiple pages are resolved, as they are independent)

"Method 1" means that you surround some `/FmX Do` with a marked content sequence that has an MCID,

So taken all together:
> /Span <</MCID 1>> BDC **/Fm5 Do** EMC
> /Span <</MCID 2>> BDC **/Fm5 Do** EMC 

is, for the purposes of rendering, equal to 

> /Span <</MCID 1>> BDC 
> **...
> (clear emphasis) Tj
> ...**
>EMC
> /Span <</MCID 2>> BDC 
> **...
> (clear emphasis) Tj
> ...**
>EMC

With the caveat that you ignore stuff about MCIDs in the substitution, 
since you have to look those up relative to the immediate containing stream.
"Method 1" has happened with the fact "/Fm5 Do" is in between a BDC and an EMC,
MCID 1 refers to the first use, and MCID 2 refers to the 2nd use.

**Comment by myang-apryse** [comment]
> Is the interpreter supposed to keep a list of all the form XObjects it's encountered so far whilst processing a page stream, and to check every new invocation against that list?

I would recommend you read the relevant sections on Streams, XObjs, etc.

The model of PDF is that the a stream contains several parts:
The "stream content", and you can consider that to just be a text block that says stuff like "/Span <</MCID 1>> BDC /Fm5 Do EMC".
Separately, it has a "Resource" Key in its dictionary, which is itself a dictionary, 
and in its own "XObject" key, lists references to the XObjects that the parent stream wants to use,
and the stream will refer to these XObjects by name in the stream content.
A page is like a "pseudo stream" wrt its "Contents" being able to be comprised of multiple objects.

So a spec conforming document's pages & streams will list (i.e. have pointers to) other streams which are the XObjs it invokes.
And when processing stream content, you just look up in the Resources by name for anything you encounter.

**Comment by myang-apryse** [comment]
> If so, an example that differentiated the 2 different cases

The thing we're forbidding in the case of object reuse is "method 2", 
which is that you can forgo having explicitly having marked content sequences, and just say the entire XObj (e.g. Fm5) is tagged.

So in the example, you would take out the BDCs & EMCs surrounding Fm5 Do, 
and just add to `5 0 obj` a "StructParent" (singular) key (see Parent Tree in spec).

Why it doesn't work is because if you follow the process,
can only end up at 1 destination structure element node for the entire Fm5 XObj, 
which means you can't differentiate between the 2 different instances of **/Fm5 Do**, 
but it's fine if you only have 1 instance (across all pages, i.e. no reuse)

**Comment by myang-apryse** [comment]
Actually, after going through the exercise of the example, I think I see what Note2 was referring to wrt 
"just a single object reference suffices to identify all of them" if they happen to be on the same page.

I think this is a bastardization of the lookup process that coincidentally works, and we should not encourage it.
I think it's just easier for everyone if there is a bijection between what I'll call "structure leaves" and (tagged) stream content.
("structure leaves" being instances in the "K" array corresponding to MCID, MCR, OBJR ways of referring to content)

If you do as Note2 says, and you use a single OBJR for obj 5, 
then if you don't care about the instance of occurrence, it works because
- in the "forward" lookup, when you go from structure to content, , you'll always be able to find the entire content subblock corresponding to obj 5, and use that for rendering/replacement in structure traversal.
- in the "backward" lookup, from content to structure, you can encounter multiple instances of Fm5 Do, and if you follow the StructParent, you'll end up at a single structure element node, which is fine iff you have exactly 1 such node as parent of the OBJR "leaves"

I propose we get rid of Note2 altogether, and just say only 1 OBJR can exist for each pdf obj in the entire structure tree?
- note that this is actually functionally more restrictive than before, but I think produces more clarity and consistency
- even if wanted to keep Note2, it should to be amended (or clarified?) by saying that only a single structure element node must contain all OBJRs referring to this XObj (I guess it's implicit via "If it is important to distinguish between multiple renditions", but at least I was confused that that meant)

--- 

Separately @sgaither,  I feel like example does not clearly illustrate the need to switch from OBJRs to MCID/MCRs, 
as the use case puts the both instances under the same structure node, for which multiple OBJRs will coincidentally not fail.
I suggest this example be kept to separately illustrate use of single ActualText replacing multiple renderings, but a separate one be added where we put the different instances of `FmX Do` under different structure nodes.
(potentially with comments indicating that "if you used OBJR here, you wouldn't be able to do the reverse lookup")

--- 

Separately separately, I suggest that we clearly emphasize (and repeat in different notes as appropriate),
that any subblock of content, once tagged, can not contain more tagging within it.
With examples showing some selection/combination of:
- <</MCID 1>> BDC /FmX Do EMC together with Fm5 being OBJR tagged (have StructParent singular) is illegal
- Fm5 being OBJR tagged, and any transitive child having any MCID or OBJR tagging is illegal
- Nesting marked content sequences (wthin the same stream) that have MCID is illegal
- However, nesting of marked content sequences that DON'T have MCID with ones that do is legal

**Comment by myang-apryse** [comment]
Related (but I'm not sure if it falls outside the scope of this ticket),
what should be the appropriate understanding of non-"Do" methods of stream/content incorporation?

For example (and where it's relevant to this OBJR discussion), 
Annotations (usually Link), can be referred to using OBJRs, can we share annotation objects across pages?

WRT to the "no tagged nesting" rule, I guess each annotation forms a separate "inclusion tree"?
But what about things like TilingPatterns? Can/Should those count as XObj invocations in some sense?

If we wish to support tagging within them, do we have to say something like if you SC a tiling pattern, 
and scn multiple times, then that's has the same restrictions as performing multiple "do" calls, and you can't use OBJRs on them.

What about things like branching AP streams? (i.e N, R, D, and annotation sub dictionaries for /On /Off)
Should those be considered as non-interfering? 
That is, if you use the same XObj for multiple of those branches, 
in any particular state, you've only used it once, so multiple OBJR is allowed?
(I guess that provides a use case that invalidates what I said about "1 OBJR can exist for each pdf obj in the entire structure tree",
 and we'll need a more complex description or go back to Note2)

**Comment by myang-apryse** [comment]
Sorry for the long-winded comments, let me know and I'd be happy to edit/split/delete sections.

**Comment by sgaither** [comment]
Well, I am not sure where this leaves us.  First, the only thing I can be clear on is Robin's first comment that object 2 in the example should correctly point to object 1 as its parent (and I am updating the example to reflect that observation - thank you, Robin).  Beyond that, I think that I am now perhaps confused on all of the above analysis and it is now unclear to me that any argument I give will convince people as to the intent of the specification (at least so far as I understand it, which now may potentially be not on as solid ground as I would like).  I need time to re-read 14.7.5 in its entirety to better understand all of its implications.  @petervwyatt, at this point I need to withdraw my original proposal, and this very long issue thread needs to remain unresolved.

**Comment by robinwatts** [comment]
> @robinwatts (I apologize if my comments repeat some stuff, I might have misunderstood your level of understanding)

No apology necessary! I like to think I understand PDF well enough at this point to know how streams and Form XObjects work :)

I am a newcomer to using the structure tags/structure tree though.

> **MCID is a number that is stream local**

D'Oh. Of course, I had forgotten that. As you say, that completely removes my worries about it being used from multiple pages.

> My understanding is that when you do the text extraction, you actually reconstruct the extraction through
> a traversal from the structure tree side, looking up pieces of the stream content as required (i.e. you skip
> stuff that's not tagged, and the stuff that is tagged is not necessarily in the same order or same # of 
> instances as in the streams).

That's certainly NOT how we do text extraction in either MuPDF or Ghostscript, and I'd be really surprised if the specification was written deliberately so as to mandate code working in this way.

My concerns here sprang from the discussion of this bug at the last TWG meeting. My understanding of @sgaither's claims was that EXAMPLE-6, *even without the ActualText* would result in 'clear emphasis' being included in the text extracted output just once.

If I have misunderstood that claim, then please excuse the noise.

If that is indeed the claim that @sgaither was making, then my concerns over how an interpreter is supposed to know to do that stand.

**Comment by stechio** [comment]
@sgaither:
> [cut] Upon more careful examination of the specification, the verbiage at the end of 14.7.5.2 is quite clear that: "A form XObject that is painted with multiple invocations of the Do operator shall be incorporated into the Document's logical structure only by the first method, with each invocation of Do individually associated with a structure element"

I'm sorry, but *what you are referring to (i.e., marked-content sequences as content items (14.7.5.2)) is NOT what sub-issue 2 was about* (i.e., **object references to PDF objects as content items** (14.7.5.3), as described in NOTE 2) — I am totally aware of the use of structural marked content for multiple invocations of the `Do` operator, but that's NOT the point on topic. For the sake of clarity, I bring here the problematic assertion from NOTE 2, as reported in [my initial comment](https://github.com/pdf-association/pdf-issues/issues/343#issue-1889163231):
> If the referenced object is rendered on multiple pages, each rendering requires a separate object reference.

The conundrum here is that NOTE 2 claims that *separate **object references to PDF objects** (NOT marked-content sequences!)  are required for each rendering in case of multiple pages*, despite [we all agree](https://github.com/pdf-association/pdf-issues/issues/343#issuecomment-2103648313) that a PDF object as content item can be associated, via structural parent tree, to ONE AND ONLY ONE parent structure element.

**Comment by myang-apryse** [comment]
@stechio
My god you're right, I had mistakenly thought they were in the same section. 
What I was referring to as "method 2 is actually "method 3" (since 2 is taken)

However, I think the overall logic is still sound and this line does offer insight into how to resolve multiple XObj invocation.
if you treat OBJR as surrounding the entire XObj1 stream (which contains the "do" for XObj2) with an MC.

So taken all together, we get that if an XObj is reused,
- we are not allowed to have tagging inside it (no method 2)
- NEW: nor are we allowed to have it be tagged via OBJR (no method 3)
- it can only be tagged via surrounding individual do operators with MC (method 1)

My arguments above on the validity of method 3 + reuse stand, and I think it's an anomaly, and we should explicitly forbid it.
(though its problem is less complex than method 2, which can have individual MCs point at different parents,
and the full consistency condition is that each individual lookup ends up at the only parent)

**Comment by myang-apryse** [comment]
@robinwatts 
My interpretation of the "semantically once using AcutalText" comes from this quote

> See Example 6 in this subclause for multiple invocation of a form XObject to overprint a phase **in conjunction with the use of ActualText to indicate that the phrase is to be read only once** ...

**Comment by robinwatts** [comment]
> @robinwatts My interpretation of the "semantically once using AcutalText" comes from this quote
> 
> > See Example 6 in this subclause for multiple invocation of a form XObject to overprint a phase **in conjunction with
> > the use of ActualText to indicate that the phrase is to be read only once** ...

Yes, the way the example is constructed, the ActualText will ensure that the text within the 2 invocations of the same XObject will only count once. I am not disagreeing with that. That seems perfectly fine and sensible.

My point of concern is with the claim that I believe was made during the meeting, which was that *even without the ActualText*, the contents of the XObject would only count once in the structured output, despite being invoked twice.

If I've misunderstood the claim, all that is required is for @sgaither to tell me so, and I'll cease barking up this particular tree, and retire with my apologies for wasting everyone's time.

If I haven't misunderstood the claim, then either 1) it's a reasonable claim and it's just my ignorance at play here, so I need to go away and try and understand how it can possibly work, OR 2) there is a genuine issue here.

**Comment by petervwyatt** [comment]
I'll also note that 14.7.5.1.1 "Content items" defines a bunch of other file format "shall" requirements. "Content item" (and "structure content item") are also used as defined terms of art with all those requirements applying  (i.e. has to be a leaf node, etc) - and which are then partially restated differently in the lower subclauses.

On a closer re-read I also noticed:

* last bullet in 14.7.5.1.1 explicitly mentions reference XObjects but I imagine this processor requirement also applies to any content that can be dragged in from an external source (e.g. any stream that uses **FFilter**). Not because you should, but because you can!

* 14.7.5.2 1st bullet states "_Although the tag associated with a marked-content sequence is not directly related to the document’s logical structure, it should be the same as the structure type of the associated structure element._" - so this begs the question how do you know what a "_marked-content sequence corresponding to a structure content_" is (phrase from 14.7.5.1.1, 2nd set of bullets)? It appears the answer is when the property list contains an **MCID** entry so this probably needs to be spelt out for cases when the marked content tag is NOT the same as the logical structure...
    - this then begs the question if an **MCID** entry can validly exist on some, none or all of the marked content tags formally defined elsewhere in ISO 32000-2! e.g. what if the tag was **OC**? **ReversedChars**? **Tx**? **AF**? etc. Current wording would imply yes - and thus both the tag requirements AND the MCID would apply. Again, not because you should, but because you can!

**Comment by petervwyatt** [comment]
Further:
* Table 358, **Pg** entry is described as _optional_, but in keeping with elsewhere it should be "_Sometimes required_" since the very last portion of the description says "_it shall be used if the structure element has no such entry_" which can only apply if it is always present for that condition. 
    - A better rewording would be "_This entry is required if the structure element containing the object reference has no **Pg** 
entry._"
    - And we normally say "_takes precedence over_" rather than "_overrides_"

**Comment by myang-apryse** [comment]
> It appears the answer is when the property list contains an MCID entry so this probably needs to be spelt out

This is spelt out under 14.7.5.2's 2nd bullet point I think? Also includes implication of stream-locality.
> The marked-content sequence shall contain a property list (see 14.6.2, "Property lists") **containing an MCID entry**, which shall be an integer marked-content identifier that uniquely identifies the marked-content sequence **within its content stream**

---

> how do you know what a "marked-content sequence corresponding to a structure content" is

I had presumed that this condition meant "if there's a structure element node at the end if you do the lookup using the containing stream's StructParents (plural) entry and the MCID". That is to say, if the chain breaks at any point, 
(e.g. StructParents not existing, ParentTree node having a null reference at this index etc)
then this is no longer an "MC corresponding to a structure content".

---

> e.g. what if the tag was OC? ReversedChars? Tx? AF? etc. Current wording would imply yes - and thus both the tag requirements AND the MCID would apply. Again, not because you should, but because you can!

It's interesting to note that you probably should separately nest the MCID MC within/around these ones,
since they can be functionally different, causing you to violate the recommendation that the tag be the same as structure element type. E.g. If you had 2 different OC, and it acts as a /P in one place, but an /LI in another.
Maybe we should just explicitly forbid these ones from being structural MC so that "you can" is no longer true?

But we might be digressing a bit too far from the original discussion on OBJRs here.

**Comment by myang-apryse** [comment]
Speaking of 14.7.5.1.1, this wording on the identity/definition of of Content Items always bothered me
(This is tangentially related to OBJR discussion since OBJR is one such method of referring to Content Items)

> Content items shall be leaf nodes of the structure tree

This conflates thing being referred to, with the actual methods of referring to them.
Yes, there is some sort of semantic equivalencebetween them, however, we're in the section of the spec defining the form of how such relations take place syntactically, so I think it's confusing/circular, and we should distinguish these 2 concepts. 
Hence why I mentioned above a "what I shall call 'structure leaves'".

Additionally, this line is playing fast and loose with the definition of tree **nodes**, and (potentially?) conflicts with definition of structure element nodes above. Since the MCID method of referring to content isn't even a dictionary, so it doesn't have the capability of having children via the K entry.

I propose we change to some kind of "ContentItems can be referred to in the structure tree in 3 ways, MCID, MCR, OBJR",
and there we can enforce a bijiection and say that if 2 of them correspond, then the lookups in both directions should find each other*.
*: well, struct **parent** look up actually finds the containing parent structure element node, which technically speaking,
may (can not should?) contain multiple references to the same content and still be consistent.

**Comment by petervwyatt** [comment]
>>  It appears the answer is when the property list contains an MCID entry so this probably needs to be spelt out

> This is spelt out under 14.7.5.2's 2nd bullet point I think? Also includes implication of stream-locality.

>>  The marked-content sequence shall contain a property list (see 14.6.2, "Property lists") containing an MCID entry, which shall be an integer marked-content identifier that uniquely identifies the marked-content sequence within its content stream

I completely agree - the issue is the ordering of things in the spec and the fact that different terminology/phrasing is being used and only sometimes identified as an italic term-of-art (not an approved ISO-ism, but the way PDF specs have always been written). I also believe all the definitions need to be file format related and not processor requirements. 

"Stream locality" as you describe is stated but subtly with the phrase "_... within its content stream..._"

**Comment by stechio** [comment]
After over a year of discussion, can someone please clarify in non-convoluted, plain English, what's the consensus about [SUB-ISSUE 2](https://github.com/pdf-association/pdf-issues/issues/343#issue-1889163231):

> What is the **correct way to place in the logical structure multiple separate object references (one per page) of the same PDF object as content item**? (EXAMPLE 1 in 14.7.5.4 (Finding structure elements from content items) only shows the simple case of a single object reference; IMO, _that example should be expanded to address also the case depicted in NOTE 2, ie multiple separate object references, one per page_ — BTW, apparently that example is the only one in the whole ISO PDF spec to mention object references, so it is crucial to integrate it)

Thanks

**Comment by petervwyatt** [comment]
@sgaither?

**Comment by sgaither** [comment]
@petervwyatt I don't have a great answer on this one.  There does not appear to be as clean a solution as I had thought, and there is absolutely no consensus SFAICT.

**Comment by stechio** [comment]
Apparently, NOTE 2 of subclause 14.7.5.3 (*multiple separate object references, one per page*) has survived unscathed since the introduction of Logical Structure (PDF 1.3, section 8.4.3) — really disappointing that the original authors didn't provide an example to demonstrate their point... I guess there's no chance anymore, after 20+ years, to reach out to Adobe and retrieve the original proceedings, or to contact its authors for some clues... Anyway, it would be bizarre if nobody in the industry stumbled upon this case during such a long time...

As pointed out in my [previous comment](https://github.com/pdf-association/pdf-issues/issues/343#issuecomment-2137791007), the conundrum here is that NOTE 2 claims that _separate **object references to PDF objects** (NOT marked-content sequences) are required for each rendering in case of multiple pages_, despite [we all agree](https://github.com/pdf-association/pdf-issues/issues/343#issuecomment-2103648313) that a PDF object as content item can be associated, via structural parent tree, to ONE AND ONLY ONE parent structure element.

**Comment by u-fischer** [comment]
> As pointed out in my https://github.com/pdf-association/pdf-issues/issues/343#issuecomment-2137791007, the conundrum here is that NOTE 2 claims that separate object references to PDF objects (NOT marked-content sequences) are required for each rendering in case of multiple pages, despite https://github.com/pdf-association/pdf-issues/issues/343#issuecomment-2103648313 that a PDF object as content item can be associated, via structural parent tree, to ONE AND ONLY ONE parent structure element.

Well personally I always wondered if there is an error in the spec and if the `StructParent <some number>` in the (e.g. link) annotation was actually meant to be `StructParents [array of numbers]`. 

I once made a picture to illustrate the use of the OBJR and from it was clear that it simply adds another object to a lookup circle:

![image](https://github.com/user-attachments/assets/c808e59e-4944-4e67-b141-2dfdf0d26015)

This looked quite senseless. What do you gain from the additional OBJR object? Looking up the objects would be possible without it. 

But if you assume that the StructParents can be an array the additional OBJR object make sense. Then you can reuse an annotation in multiple structures without violating the principle that every object is only the kid of one structure element:

![image](https://github.com/user-attachments/assets/85807186-3b61-4889-bf41-a7cc724c7900)

**Comment by stechio** [comment]
@u-fischer Eureka! That's precisely what I expected from the spec and I was hopelessly looking for, brlliant!

Just a minor observation: switching to `StructParents` entry would be semantically incorrect since, as you know, that entry is already allocated for another structural parent retrieval sub-mechanism (marked-content sequences within content stream): I would stick to existing `StructParent`, extending its type to support also `array` as an alternative to current `integer`.

**Comment by car222222** [comment]
@stechio Sounds sensible!

**Comment by u-fischer** [comment]
> Just a minor observation: switching to StructParents entry would be semantically incorrect since, as you know, that entry is already allocated for another structural parent retrieval sub-mechanism (marked-content sequences within content stream): 

Well perhaps `StructParents` in an annotation could do/mean something different to the other `StructParents` but I don't care much about the name, I merely put an `s` at the end because it looked more grammarly correct ;-). So `StructParent [array]` is completly fine with me.

**Comment by stechio** [comment]
@u-fischer:

> @stechio
> 
> > Just a minor observation: switching to StructParents entry would be semantically incorrect since, as you know, that entry is already allocated for another structural parent retrieval sub-mechanism (marked-content sequences within content stream):
> 
> Well perhaps `StructParents` in an annotation could do/mean something different to the other `StructParents` but I don't care much about the name, I merely put an `s` at the end because it looked more grammarly correct ;-). So `StructParent [array]` is completly fine with me.

Table 359 (Additional dictionary entries for structure element access) dictates that *ALL objects that are structural content items (i.e., form or image XObjects, or annotations) are REQUIRED to use `StructParent`*, so there is no leeway here, we have to honor its name :wink:

**Comment by mrbhardy** [comment]
The PDF Reuse TWG proposes that we resolve the understanding to the three use cases:

1. Do operator in content stream wrapped in MCID (page structparents reference)
2. Single reference to OBJR using back pointer in StructParent entry
3. Multiple references to internal MCIDs using MCR/STM object and the StructParents back pointer to the array of structure elements referencing content within the object.

Matthew and Shawn commit to creating an example of number 3 to include in the spec via an errata.

**Comment by mrbhardy** [comment]
Addendum: An application note would be helpful to explain this situation more fully without cluttering the specification.  Matthew commit to write such a note and review with the Reuse TWG before distributing more generally.

**Comment by stechio** [comment]
@mrbhardy:

> The PDF Reuse TWG proposes that we resolve the understanding to the three use cases:
> 
>   1. Do operator in content stream wrapped in MCID (page structparents reference)
>   2. Single reference to OBJR using back pointer in StructParent entry
>   3. Multiple references to internal MCIDs using MCR/STM object and the StructParents back pointer to the array of structure elements referencing content within the object.

¿Can we deduce that the following, contradictory statement in NOTE 2 of subclause 14.7.5.3 (PDF objects as content items), referenced by [sub-issue 2](https://github.com/pdf-association/pdf-issues/issues/343#issue-1889163231), is therefore incorrect and requires to be fixed/removed:
 
> If the referenced object is rendered on multiple pages, each rendering requires a separate object reference.

**Comment by myang-apryse** [comment]
Just a comment on 3. in that perhaps we should discourage it,
due to a limitation that we should also explicitly state:
- that, even though the xobj stream itself can be reused, any MC may only have a single struct parent
- that is to say, you cannot have 2 MCRs w/ 2 different parents referencing the same piece of MC
- which means, that parts of the shared obj must be considered UNTAGGED at each shared location
  - which would make it impossible to satisfy certain extension specs such as PDF/UA, requiring all content to be tagged.

example: I have a shared x obj with MCs 1, 2, 3, 4
- I put in on pg 1 & 2 (or 2x on the same page, each instance being different).
- I can have (pg1, mc1), (pg2, mc2), (pg1, mc3), (pg4, mc4), with each mc being tagged once, but on different pages
- but I cannot have MCR(pg1, mc1), (pg2, mc1) ..., because there's no way perform the lookup via StructParents for mc1 for the xobj

**Comment by myang-apryse** [comment]
@u-fischer 
Wrt to the array idea for multiple reference, we should think this through very carefully,
because it may add a lot of undesirable complexity in the general case.

Conclusion: If we do it, I suggest we do it only for OBJRs for Annotations.
Rationale:
- I've recently come up against a use case where this would be useful, 
  and that's if a text-markup annotation covers disparate pieces of content
  - e.g a hyperlink that covers 2 cells of a table.
- because Annotations cannot be referred to by MCRs, we must use OBJRs
- in order to achieve multiple OBJRs, we need StructParent (singular) to be an array

Why not do it in general:
- this is a very niche use case, and we do not want to force checks whether the value is a Number or an Array all the time
  restricting this to only Annotations gives a clear condition for which we need to take this into account.
- if we want this to work on MCRs, we'd need a 2d array, and that's way too much work

Conversely, 
maybe we offer explicit guidance that you should simply split your annotation into 2,
in the same way that you have to for hyperlinks that span across multiple lines,
s.t. each annotation only needs to be under a single structure element?

**Comment by myang-apryse** [comment]
Actually, I keep forgetting about this
> ... A form XObject that is painted with multiple invocations of the Do operator shall be incorporated into the Document's logical structure only by the first method...

Taken strictly, it actually just forbids
> 3. Multiple references to internal MCIDs using MCR/STM object and the StructParents back pointer to the array of structure elements referencing content within the object.

Due to the fact that
> So taken all together, we get that **if an XObj is reused,**
> - **we are not allowed to have tagging inside it (no method 2)**
> - NEW: nor are we allowed to have it be tagged via OBJR (no method 3)
> - it can only be tagged via surrounding individual do operators with MC (method 1)

So we're actually only left with 
1. Multiple-ref XObjs can only be tagged via on of its ancestor's MC, surrounding a DO operator
2. Single-ref XObjs (and Annotations/AP streams) can additionally have the option of using OBJR
    - If we choose to, we can add the extension/exception I suggested above for multi-ref annots,
    - Otherwise, the corollary is that Annotations/AP streams can only be referred to once, as there is no ancestor DO

**Comment by car222222** [comment]
Is this still unresolved?

**Comment by myang-apryse** [comment]
I think there's a deeper unresolved problem with the tagging spec itself (wrt to shared streams/objects),
and that this OBJR thing is just a manifestation of it.

I have some musings, but it's a bit long winded, should I make a separate ticket and link it?

The TLDR (without context), is that 
**tagging scope must be _strictly_ above sharing scope**, otherwise the parent tree lookup is ill defined.
This restriction is not well explained in the existing spec, and may even have omissions/contradictions, such as this OBJR thing.

**Comment by petervwyatt** [comment]
@myang-apryse - I think a separate issue might be best... just mention this one

**Comment by myang-apryse** [comment]
@petervwyatt I've created https://github.com/pdf-association/pdf-issues/issues/544
